_posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown

```diff
@@ -0,0 +1,287 @@
+
+The whole authentication flow consists of a few steps.
+1. The GKE MDS requests OpenID Connect (OIDC) signed JSON Web Tokens (JWT) from the Kubernetes API server.
+1. The GKE MDS then requests an access token for the K8s Service Account.
+1. IAM validates the bindings and OIDC and returns an access token to the MDS.
+1. That access token will be sent back to IAM, and a short-lived GCP service account token will be issued.
+1. It is then returned to the GKE MDS and passed on to the workload.
+
+Finally, the pod can use the short-lived service account token to access GCP resources.
+See Figure 4.
+
+<p align="center">
+  <img alt="workload-identity-level-3" title="Workload Identity Level 3" src="assets/images/2022_03_level_3.png" style="border:1px solid #2C4D56">
+</p>
+
+<p align="center">
+  <b style="font-size:0.7vw" >Figure 4. Workload Identity.</b>
+</p>
+
+## [Practical Example In Terraform]
+
+What good is theory without the practical part?
+
+For that, I will be using Terraform and GCP to show the GKE Workload Identity in action.
+
+From this point on, I will assume that the reader has the following:
+[Terraform 1.1 or older] and gcloud CLI installed. Functional GCP credentials (hopefully not configured via a User-managed service account key), authorized for creating and managing GCP projects. Including all of its underlying resources.
+
+I've made a [Terraform root module] that creates a:
+- GCP project (it can also reside under an Organization folder).
+- VPC with private subnets.
+- Private GKE Cluster with Workload Identity enabled.
+- Node pool with a Workload Identity enabled in `GKE_METADATA` mode. [Google recommends] including this attribute, so node creation fails if Workload Identity is not enabled on the cluster.
+- A shell script named `k8s.sh` for the creation and configuration of resources inside the GKE cluster.
+
+See Figure 5.
+
+<p align="center">
+  <img alt="workload-identity-level-2.5" title="Workload Identity Level 2.5" src="assets/images/2022_03_level_2.5.png" style="border:1px solid #2C4D56">
+</p>
+
+<p align="center">
+  <b style="font-size:0.7vw" >Figure 5. Simplified representation of the cloud resources created by the Terraform root module.</b>
+</p>
+
+For more details about the module and its configuration, refer to the [README.md].
+
+I would have preferred to create all of the resources in one go, but unfortunately, it is not possible with Terraform’s CRUD cycle. The problem is that the K8s provider cannot authenticate against the GKE cluster in one apply run. That makes perfect sense because the GKE cluster is not present in the application stage, and the K8s provider cannot authenticate.
+To programmatically create all of the resources, I utilize the [local_file] provider to generate a helper bash script from a Terraform template. The bash script `k8s.sh` will appear in the working directory of the root module after a successful apply run. It can be [overiden to any valid path] in the Operating System.
```

Comment:
```suggestion
To programmatically create all of the resources, I use the [local_file] provider to generate a helper bash script from a Terraform template. The bash script `k8s.sh` will appear in the working directory of the root module after a successful apply run. It can be [overirden to any valid path] in the Operating System.
```

Reply:
I will just change utilize to use, the other word looks like a unintentional typo.